### PR TITLE
ME-1731: create border0 terraform provider

### PR DESCRIPTION
**Pull request by waltzofpearls**
- setup terraform provider skeleton: `main.go`, `border0/provider.go`
  and `border0/resource_socket.go`
- add docs generate tool: `tools.go` and `make docs` command
- add examples for terraform docs: `examples/provider/provider.tf` and
  `examples/resources/border0_socket/resource.tf`
- add more examples for testing: `examples/basic` and `examples/development`
- setup terraform release and publish to registry: `.goreleaser.yml`,
  `terraform-registry-manifest.json` and `.github/workflows/release.yml`
- add generated docs

### CODE REVIEW

File: border0/provider.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	border0client "github.com/borderzero/border0-go/client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// ProviderOption is a function that can be passed to `Provider()` to configures it.
+type ProviderOption func(p *schema.Provider)
+
+// Provider returns a Border0 implementation and definition of terraform `schema.Provider`.
+func Provider(options ...ProviderOption) *schema.Provider {
```
Comment (adrianosela) [inline border0/provider.go:16]: This is awesome

File: border0/resource_socket.go [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+				Sensitive:   true,
+				Description: "The upstream username.",
+			},
+			"upstream_password": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Sensitive:   true,
+				Description: "The upstream password.",
+			},
+			"upstream_private_key": {
```
Comment (adrianosela) [inline border0/resource_socket.go:115]: Still not a fan of this flat structure...

File: border0/resource_socket.go [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+		socket.Description = v.(string)
+	}
+
+	if v, ok := d.GetOk("upstream_type"); ok {
+		socket.UpstreamType = v.(string)
+	} else {
+		// if upstream type is not set, use socket type as default upstream type
+		// except for database sockets, which use mysql as default upstream type
+		switch socket.SocketType {
+		case "http", "ssh", "tls":
```
Comment (adrianosela) [inline border0/resource_socket.go:288]: The go sdk should have consts for these

File: examples/advanced/border0_connector.tf [hunk 4]
```diff
@@ -0,0 +1,142 @@
+// ===========================================================================
+// Create a connector with plugins
+// ===========================================================================
+
+resource "border0_connector" "my_connector" {
+  name = "my-connector"
+  description = "my connector"
+
+  aws_ec2_plugin {
```
Comment (adrianosela) [inline examples/advanced/border0_connector.tf:9]: Plugins should not be prt of the connector config as they are standalone entities. 

This plugin also should be called aws_ec2_discovery_plugin -- as we may one day support non-discovery plugins

**Review by adrianosela — APPROVED** [review]

File: border0/resource_socket.go (hunk 2 above)
Comment (waltzofpearls) [reply border0/resource_socket.go:115]: Yeah, I think we should have some nested schema. I have started the work, but I removed the code to keep this PR smaller.

File: border0/resource_socket.go (hunk 3 above)
Comment (waltzofpearls) [reply border0/resource_socket.go:288]: Sounds good. I will add those.

**Pull request merged**